### PR TITLE
Added a unit test and removed unreachable code.  The while loop befor…

### PR DESCRIPTION
**Pull request by daveb1040**
…e it will only ever let index get as high as name.lenth - 1.
I wanted to add Lower_Camel_Case FieldNamingPolicy because of an issue I had the other day at work, but when I started to refactor the code, I saw a line that I couldn't get to with a unit test, so I figured I'd just add the unit test, and remove the line.
Also, I'm new to git and github, and wanted to try making a contribution to a great package.  The is my first pull request, so please bear with eh newbie if I'm not following some sort of protocol.
Thanks
David

### CODE REVIEW

**Comment by daveb1040** [comment]
I signed it!
Thanks!
David Betancourt
GitHub id: daved1040

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package com.google.gson;
```
Comment (swankjesse) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: nit: copy & paste google copyright

Comment (swankjesse) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: or Gson Authors copyright

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import org.junit.Test;
+import org.junit.runners.Parameterized.Parameters;
+
+import static org.junit.Assert.assertEquals;
+
+public class FieldNamingPolicyTest {
+	
+
+	@Test
+	public void testUpperCaseFirstLetterNonUpperCaseLetter() {
```
Comment (swankjesse) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: nit: please format this to match the rest of the project

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	@Test
+	public void testUpperCaseBlank() {
+		verifyUpperCaseFirstLetter(" ", " ");
+	}
+
+	/**
+	 * Verifies the upperCaseFirstLetter is working as expected
+	 * @param fieldName
```
Comment (swankjesse) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: you can delete this doc

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	@Test
+	public void testUpperCaseFirstJustUnderscore() {
+		verifyUpperCaseFirstLetter("_", "_");
+	}
+
+	@Test
+	public void testUpperCaseBlank() {
+		verifyUpperCaseFirstLetter(" ", " ");
```
Comment (swankjesse) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: you can probably collapse all of these tests into a single test method. It’s extremely unlikely we’ll ever want to run these tests independently, or that they’ll fail independently. That way if/when we ever want to add tests for more policies, we don’t need to write thousands of lines of code.

File: gson/src/main/java/com/google/gson/FieldNamingPolicy.java [hunk 5]
```diff
@@ -149,10 +150,6 @@ static String upperCaseFirstLetter(String name) {
       firstCharacter = name.charAt(++index);
     }
 
-    if (index == name.length()) {
```
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/FieldNamingPolicy.java]: I’m anxious that this will break compatibility with code already deployed. At this point that’s much more important than actual correctness, which is both depressing and true.

Comment (daveb1040) [inline gson/src/main/java/com/google/gson/FieldNamingPolicy.java]: I wanted to leverage this behavior for my LOWER_CAMEL_CASE fieldNamingPolicy, so before refactoring, I surrounded it with unit tests.  (I'm a unit test junkie)  I did an emma code coverage test, and realized it was impossible for me to put anything into any of my unit tests.  (The pit mutation test also flagged it as unreadable with my existing permutations) 

That said, understandable on being concerned with such a highly used class, so if you want, I'll put it back in there.  (Although, I was hoping my unit tests could mitigate the risk).  Let me know how you want to proceed, and I'll commit those changes.

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java (hunk 1 above)
Comment (daveb1040) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: These will be added in the next commit

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java (hunk 2 above)
Comment (daveb1040) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: Just installed Mars....  auto formatting wasn't set.  I pulled the formatted in from lib.  that will be in, in the next commit.

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java (hunk 4 above)
Comment (daveb1040) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: I'll make parameterized tests for these, and commit those.  The reason I don't stack them all in one is because when I do put in some code which breaks the tests, I find it desirable to have the summary of what I broke, where as, if they're all in one method, once 1 assertion fails, the test executes.  So, did I do something that broke 10 of 15 scenarios, or just 1.  I'll get this in, in the next commit

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java (hunk 3 above)
Comment (daveb1040) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: Will do

**Comment by daveb1040** [comment]
I have addressed the peer review comments.  The only thing pending is whether or not to drop the if statement.  Thanks!
David

File: gson/src/main/java/com/google/gson/FieldNamingPolicy.java [hunk 6]
```diff
@@ -20,10 +20,11 @@
 import java.util.Locale;
 
 /**
- * An enumeration that defines a few standard naming conventions for JSON field names.
- * This enumeration should be used in conjunction with {@link com.google.gson.GsonBuilder}
```
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/FieldNamingPolicy.java]: why reformat all this code? Please revert

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import junitparams.Parameters;
+
+/**
+ * Unit test for the {@link FieldNamingPolicy} class
+ * 
+ * @author David Betancourt
+ *
+ */
+@RunWith(JUnitParamsRunner.class)
+public class FieldNamingPolicyTest {
```
Comment (swankjesse) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: Dig.

File: gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+  private Object[] getUpperCaseFirstLetterParams() {
+    return new Object[] { getUpperCaseFirstLetterParamArgs("someFieldName", "SomeFieldName"),
```
Comment (swankjesse) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: might be more legible if every call to `getUpperCaseFirstLetterParamArgs` was on its own line?

Comment (daveb1040) [inline gson/src/test/java/com/google/gson/FieldNamingPolicyTest.java]: I imported the eclipse code formatter that was in the project, and had it autoformat on save...  I assumed it was required.  I'll turn it off, put that on it's own line, and commit tonight.

File: gson/src/main/java/com/google/gson/FieldNamingPolicy.java (hunk 6 above)
Comment (daveb1040) [inline gson/src/main/java/com/google/gson/FieldNamingPolicy.java]: I imported the eclipse code formatter that was in the project, and had it autoformat on save...  I assumed it was required.  I'll turn it off and revert this.  Thanks

**Comment by daveb1040** [comment]
OK...  the should do it.  Thanks
David

**Comment by daveb1040** [comment]
Are there any other issues I should address?
Thanks!
David

File: gson/pom.xml [hunk 9]
```diff
@@ -54,7 +54,13 @@
     <dependency>
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>3.8.2</version>
+      <version>4.12</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>pl.pragmatists</groupId>
```
Comment (swankjesse) [inline gson/pom.xml]: Any way you can implement this without adding another 3rd party dependency? We’ve worked hard to limit our dependencies. Gson is often embedded in other software & adding dependencies is a huge burden downstream.

File: gson/src/main/java/com/google/gson/FieldNamingPolicy.java (hunk 6 above)
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/FieldNamingPolicy.java]: This file still has tons of unnecessary changes. Please really revert.

Comment (swankjesse) [inline gson/src/main/java/com/google/gson/FieldNamingPolicy.java]: (the GitHub diff on this file should be nothing)

File: gson/pom.xml (hunk 9 above)
Comment (swankjesse) [inline gson/pom.xml]: I think you can accomplish something very similar to this with these classes:

```
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameter;
```

**Comment by swankjesse** [comment]
A few more comments. Plus you need to rebase and squash. See http://rebaseandsqua.sh for tips.

**Comment by eamonnmcmanus** [comment]
The most recent comments, from February 2016, went unaddressed so I think we can consider this obsolete.
